src/main/java/ch/njol/skript/expressions/ExprBeaconValues.java

```diff
@@ -0,0 +1,199 @@
+public class ExprBeaconValues extends PropertyExpression<Beacon, Object> {
+
+	private static boolean PAPER_LOADED = false;
+
+	static {
+		if (Skript.methodExists(Beacon.class, "getTier")) {
+			PAPER_LOADED = true;
+		}
+
+		Skript.registerExpression(ExprBeaconValues.class, Object.class, ExpressionType.PROPERTY,
+			"%-beacon%['s] primary effect",
+			"primary effect of %-beacon%",
+			"%-beacon%['s] secondary effect",
+			"secondary effect of %-beacon%",
+			"%-beacon%['s] range",
+			"range of %-beacon%",
+			"%-beacon%['s] tier",
+			"tier of %-beacon%"
+		);
+	}
+
+	private int pattern;
+	private boolean isEffect, isRange, isTier, isPrimary, isSecondary;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+		setExpr((Expression<Beacon>) exprs[0]);
+		if (matchedPattern <= 3) {
+			isEffect = true;
+			if (matchedPattern <= 1) {
+				isPrimary = true;
+			} else {
+				isSecondary = true;
+			}
+		} else if (matchedPattern <= 5) {
+			if (!PAPER_LOADED) {
+				Skript.error("This can only be used on Paper.");
+				return false;
+			}
+			isRange = true;
+		} else {
+			isTier = true;
+		}
+		pattern = matchedPattern;
+		return true;
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event, Beacon[] source) {
+		return get(source, new Getter<Object, Beacon>() {
+			@Override
+			public @Nullable Object get(final Beacon beacon) throws IllegalStateException {
+				switch (pattern) {
+					case 0,1 -> {
+						if (beacon.getPrimaryEffect() != null)
+							return beacon.getPrimaryEffect().getType();
+					}
+					case 2,3 -> {
+						if (beacon.getSecondaryEffect() != null) {
+							return beacon.getSecondaryEffect().getType();
+						}
+                    }
+					case 4,5 -> {return beacon.getEffectRange();}
+					case 6,7 -> {return beacon.getTier();}
+                };
+				return null;
+			}
+		});
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.REMOVE_ALL || isTier) {
+			return null;
+		}
+		if (isEffect && (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE)) {
+			return null;
+		}
+		if (isEffect && !PAPER_LOADED && (mode == ChangeMode.RESET || mode == ChangeMode.DELETE)) {
+			Skript.error("You can only clear 'primary' and 'secondary' effects of a beacon on Paper.");
+			return null;
+		}
+		return CollectionUtils.array(Object.class);
+	}
+
+	@Override
+	public void change(final Event event, final @Nullable Object[] delta, final ChangeMode mode) {
+		PotionEffectType providedEffect = null;
+		double providedRange = 0;
+		if (delta != null && delta[0] != null) {
+			if (isEffect && delta[0] instanceof PotionEffectType potionEffectType) {
+				providedEffect = potionEffectType;
+			} else if (isRange) {
+				providedRange = (double) ((long) delta[0]);
+			}
+		}
+        assert getExpr() != null;
+        Beacon beacon = getExpr().getSingle(event);
+		assert beacon != null;
```

Comment:
why is this indented? should be replaced with if beacon == null return null or something since this doesn't do anything in prod